Add calibration functions to the `phasor` module

bruno-pannunzio opened the pull request:
## Description
This PR adds functionality to perform calibration and other transformations to the phasor coordinates.

This intends to cover the issue #27 

## Release note

Summarize the changes in the code block below to be included in the
[release notes](https://www.phasorpy.org/stable/release.html):

```
Implementation of functions to perform calibration of phasor coordinates.
```

## Checklist

- [x] The pull request title, summary, and description are concise.
- [x] Related issues are linked in the description.
- [x] New dependencies are explained.
- [x] The source code and documentation can be distributed under the [MIT license](https://www.phasorpy.org/stable/license.html).
- [x] The source code adheres to [code standards](https://www.phasorpy.org/stable/contributing.html#code-standards).
- [x] New classes, functions, and features are thoroughly [tested](https://www.phasorpy.org/stable/contributing.html#tests).
- [x] New, user-facing classes, functions, and features are [documented](https://www.phasorpy.org/stable/contributing.html#documentation).
- [ ] New features are covered in tutorials.
- [x] No files other than source code, documentation, and project settings are added to the repository.

bruno-pannunzio commented:
@cgohlke I made this PR so that we can start  discussing handling the FLIM module as object oriented. I didn't add the tests, and also the documentation because I need the phasor function from the previous PR from Bruno to be merged. But in the meantime we can discuss this implementation and when that is pushed I will add the tests and all the documentation as stated in the contributors guide.

cgohlke commented:
I think the `FlimData` class is too overloaded and not really necessary. How about a simple function, for example, `reference_phasor` or `correct_phasor`, which takes phasor coordinates and correction phase and modulation (or correction phasor coordinates) and returns the corrected phasor coordinates?

Other functions can be added to calculate correction coordinates from measured and known coordinates, both phasor coordinates and phase&modulation. Coordinates for recalibration also fit in that scheme. Known coordinates can be calculated from single or distributions of known lifetimes and frequencies. A collection of `center_function`s would be nice.

Ideally, the functions should work with many input types (scalars, lists, numpy array, xarray) and take advantage of numpy broadcasting. For example, the phasor coordinates may include a frequency axis, or the correction values may depend on spatial coordinates.

Testing this module should not need to depend on the phasor function or data files. The tests can use synthetic input values with known results.

I don't think there is a need for a `calibrate_multiple_flim` function. It looks redundant to me. Let the user decide whether to use loops, list comprehension, map functions, dask, or other methods.

The `TAU_REFERENCES` data are nice to have, although the values depend on many environmental factors. Ultimately the user must provide accurate correction/reference coordinates from their measurements. Did you check the copyright on the ISS website?

cgohlke commented on src/phasorpy/flim.py:
```diff
@@ -0,0 +1,498 @@
+                   center_function=None, **kwargs)
+        Calibrates the FlimData object based on reference data and correction parameters.
+
+    Attributes
```
Public class attributes should be typed and can be documented by inline docstrings I think. That avoids duplication.

cgohlke commented on src/phasorpy/flim.py:
```diff
@@ -0,0 +1,498 @@
+        ValueError
+            If the original data is not provided and either the average phasor or the real and imaginary coordinates are also not provided.
+
+        Examples
```
Examples can go into the class docstring.

bruno-pannunzio commented:
@cgohlke thanks for all the comments. I agree with you that the FlimData class may have been overloaded and difficult to be useful to everyone. I implemented the functions to get calibration parameters, calibrate phasor data and some methods to calculate center of mass. This are some that didn't needed new dependencies, so thats way there can be more but we need to discuss adding dependencies like scipy or sklearn. We can think of adding new methods later like kmeans.

The calibration function now can be used with arrays, sequence type like lists, and scalars that are converted to float. The function now returns a tuple with float if the input was scalar, and a tuple with numpy arrays otherwise. Do you think it would be necessary for the output to be the same as the input, for example if the input data were lists, it should return lists? Or is it okay to return numpy arrays?

cgohlke commented:
I think this revision looks much cleaner and tests are passing!

> we need to discuss adding dependencies like scipy or sklearn.

Using `scipy` or `sklearn` should be OK since we were [planning to use them later](https://github.com/phasorpy/phasorpy/blob/6614090d13e91ff407d4c4db45aa6aa4d8a65b01/pyproject.toml#L15-L26) for the ML anyway. These are big dependencies, so I would prefer to defer the import to when they are actually used unless they become used pervasively.

If you plan to add many more methods to `center_of_mass`, the function will get too large. Maybe better to factor out the methods to private helper functions and dispatch in the main function?

> The function now returns a tuple with float if the input was scalar, and a tuple with numpy arrays otherwise. Do you think it would be necessary for the output to be the same as the input, for example if the input data were lists, it should return lists? Or is it okay to return numpy arrays?

It gets really complicated to type return values based on parameter types (overloading). It's simpler to provide one function to handle arrays/lists and a separate function that takes and returns scalars.

At this point, we should discuss and agree on naming common parameters and results. Instead of `real_phasor` and `imaginary_phasor` I would prefer the shorter `real` and `imag`. Instead of `phi` and `modulation`, either `phi` and `mod` or `phase` and `modulation`. `tau` could also be named `lifetime`. Parameters should be marked positional-only, keyword-only, or none of both.

I think the current functionality of this module covers a common use case but is too limited for a library. For example, the `calibration_parameters` function only works with a monoexponential lifetime and a single frequency. What about users that have multi-frequency data or calibrated with a multi-exponential reference? Similarly, the `calibrate_phasor` function does not allow for varying calibration parameters based on spatial coordinates or frequency axis. We should be able to support this with numpy broadcasting. If it is too complicated, for example, to support multiple frequencies in one function, a set of specialized functions should be provided.

Anyway, all of this (naming, overloading, supported use cases) is up for discussion.

I get some warnings from Sphinx:
```
docs\api/flim.rst:2: WARNING: Title underline too short.

phasorpy.flim
-----------
src\phasorpy\flim.py:docstring of phasorpy.flim.calibration_parameters:16: WARNING: Definition list ends without a blank line; unexpected unindent.
src\phasorpy\flim.py:docstring of phasorpy.flim.center_of_mass:10: WARNING: Definition list ends without a blank line; unexpected unindent.
```

bruno-pannunzio commented on src/phasorpy/flim.py:
```diff
@@ -0,0 +1,238 @@
+
+The ``phasorpy.flim`` module provides functions for the calibration of time-
+resolved fluorescence lifetime imaging (FLIM) data using phasor techniques. The
+module is designed to facilitate the processing and interpretation of FLIM data,
```
Sorry I missed that line. Do you have a tool to automatically format the docstring elements to limit them to 79 characters? Because black won't cover docstrings and I tried blackdoc and docformatter but to no avail, in this case at least. Should we check all the docstring lines manually to be below the 79 characters limit?

cgohlke replied:
`flake8` caught this. Anyway, there are cases where docstrings may exceed the maximum line lengths, for example, for long URLs (https://github.com/phasorpy/phasorpy/blob/6614090d13e91ff407d4c4db45aa6aa4d8a65b01/src/phasorpy/datasets.py#L14C1-L14C108). Writing numpy docstrings is tedious. Can you try [ruff](https://docs.astral.sh/ruff/faq/#does-ruff-support-numpy-or-google-style-docstrings). Ultimately we would like to replace all linters (isort, black, pylint...) with ruff.

bruno-pannunzio replied:
I tried ruff, but it didn't catch those issues. Now with flake8 I tried in worked perfectly to check for lines that exceeded the limit. Maybe we can think of implementing some test to enforce this or add the check in the contributors guide.

bruno-pannunzio commented on src/phasorpy/flim.py:
```diff
@@ -0,0 +1,238 @@
+def calibrate_phasor(
+    real_phasor: ArrayLike | Sequence | float,
+    imaginary_phasor: ArrayLike | Sequence | float,
+    calibration_params: tuple[float, float],
```
I agree actually, I initially implemented it that way but then thought that as calibration_parameters return a touple, it would be in terms of the pipeline that what the other function returns can be directly used here as the parameter arguments. But maybe it is of no use, so I will go back to the two parameter argument for phase and modulation correction. 

Sorry I didn't understand this part regardint the correction phasor coordinates:

> Or, since the function transforms phasor coordinates, accept correction phasor coordinates instead? 
>

cgohlke replied:
Since the function is named after "phasor" and the input are phasor coordinates, I thought it would make sense for it to also accept phasor coordinates (in addition to or instead of phase and modulation) for correction. In the end it is just a simple coordinate transformation (cartesian->polar)...

Regarding tuple vs separate parameters, users can pass the `calibration_parameters` results by unpacking `calibrate_phasor(real, imag, *calibration_parameters(...))`.

bruno-pannunzio replied:
I get it now, thanks for the clarification. I think this is a very good idea, since for example R64 files contain the modulation and phase. I think I can implement it in a way so that it is not necessary to transform to cartesian and then back to polar, because it can be worked directly in the polar coordinates and be more efficent.

cgohlke replied:
Yes, the correction is done in polar space. It would just be an convenience to alternatively accept phasor coordinates.

cgohlke commented:
Some other thoughts: 

- The `calibrate_phasor` and `center_of_mass` functions are not specific to FLIM and should probably be moved to the `phasor` module. `calibrate_phasor` could be useful to transform (scale and rotate) spectral phasor coordinates.
- `laser_frequency` is time-domain specific and should better be named `frequency` to include frequency-domain data.

bruno-pannunzio commented:
I agree. I will move `calibrate_phasor` and `center_of_mass` to phasor module. Maybe after the previous PR from Bruno is merged to avoid conflicts. 

Also, `center_of_mass` is not strictly related to phasor manipulation and can be used for other purposes I think. Do you think we should create another module for `tools` or similar, or should we keep it in the phasor module. I think other functions for processing may have the same destination, for example functions to filter the data before plotting and thresholding.

cgohlke commented:
> I agree. I will move calibrate_phasor and center_of_mass to phasor module. Maybe after the previous PR from Bruno is merged to avoid conflicts.

@schutyb ? 

Merge conflicts can be resolved. It's just extra work...

> Also, center_of_mass is not strictly related to phasor manipulation and can be used for other purposes I think.

Do you have an example? It's OK to introduce a `tools` module. At this point functions can still be moved around anyway.

bruno-pannunzio commented:
@cgohlke I made a few changes to take into account what we have been discussing. I tried to modify the functions to accept both cartesian and polar data coordinates. I don't know if it is the best way to go to try for the same function to handle both type of data or to create two different functions depending on the data. I made arguments optional for passing data, maybe the best way to go rather than what I did is to add one argument that may adopt 'cartesian' or 'polar' and handle the data accordingly, instead of 'real', 'imag', 'phase' and 'modulation'.

Same for the `center_of_mass` function that I divided it into two, but maybe as we add more methods, as you said, it would be better to factor to private methods and dispatch, while handling if it is cartesian or polar data, as to avoid convertion from one to the other type.

I still have to implement many tests to consider the new implementations for polar data, but I wanted to check if this was the correct way or if it's better to do something else before proceeding.

cgohlke commented:
I think we better take a step back and first agree on the terminology, standardized names in code, function signatures, and use cases before continuing discussing the implementation. I'll try to put together a draft for Monday's meeting.

cgohlke commented:
I suggest to break down the calibration/reference functionality into several basic functions to support more advanced use cases and make testing an reviewing easier. 

Here's a simple case (in pseudo code):
```Python
signal = read_format('signal.format')
reference_signal = read_format('reference.format')
reference_lifetime = 4.0  # may be extracted from `reference_signal.attrs`
frequency = 80.0  # may be extracted from `signal.attrs`
axis = 0  # may be determined from `signal.attrs`

phasor = phasor_referenced(
    *phasor_from_signal(signal, axis=axis)[:2],
    *phasor_from_reference(
        *phasor_center(
            *phasor_from_signal(reference_signal, axis=axis)[:2], method='mean'
        ),
        *phasor_from_lifetime(frequency, reference_lifetime)
    )
)
```
The function signatures could look like this:
```Python
def phasor_referenced(
    real: ArrayLike,
    imag: ArrayLike,
    real0_or_phase0: ArrayLike = 1.0,
    imag0_or_modulation0: ArrayLike = 0.0,
    /,
    *,
    is_polar = False,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return referenced/calibrated phasor coordinates."""
    if is_polar:
        phi0, mod0 = real0_or_phase0, imag0_or_modulation0
    else:
        phi0, mod0 = phasor_to_polar(real0_or_phase0, imag0_or_modulation0)
    if phi0 == 0 and mod0 == 1:
        return real, imag
    ...
    return real, imag


def phasor_from_signal(
    signal,
    /,
    *,
    harmonic: int = 1,
    axis: int = -1,
) -> tuple[NDArray[Any], NDArray[Any], NDArray[Any]]:
    """Return phasor coordinates and average intensity from signal."""
    ...
    return real, imag, fdc

def phasor_from_reference(
    measured_real: ArrayLike,
    measured_imag: ArrayLike,
    known_real: ArrayLike,
    known_imag: ArrayLike,
    /,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return phasor coordinates from reference measurement."""
    ...
    return real0, imag0


def phasor_to_polar(
    real: ArrayLike,
    imag: ArrayLike,
    /,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return polar coordinates from phasor coordinates."""
    ...
    return phase, modulation


def phasor_center(
    real: ArrayLike,
    imag: ArrayLike,
    /,
    *,
    skip_axes: tuple[int, ...] | None = None,
    method: Literal['mean', 'spatial_median', 'geometric_median'] = 'mean',
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return center of phasor coordinates."""
    return {
        'mean': _mean,
        'spatial_median': _spatial_median,
        'geometric_median': _geometric_median,
    }[method](real, imag, skip_axes=skip_axes)


def phasor_from_lifetime(
    frequency: ArrayLike,
    lifetime: ArrayLike,
    /,
    fraction_or_amplitude: ArrayLike = 1.0,
    *,
    is_amplitude: bool = False,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return phasor coordinates from lifetime distribution."""
    ...
    return real, imag
```

Obviously, I have not implemented or tested this...

If carefully implemented, this should (for example) allow calibrating with

- a reference with multiple lifetime components (provide lifetime sequence and second argument to `phasor_from_lifetime`)
- phasor coordinates extracted from the analysis of the phasor histogram (replace `phasor_center`)
- phasor coordinates varying along channel, frequency, or spatial coordinates

As for naming functions and parameters, here's a list of suggestions for common terms:

- `signal`: Time-domain, frequency-domain, or spectral data
    - `decay`, `histogram`: time-domain signal
    - `wave`: Frequency-domain signal
    - `spectrum`: Spectral signal
    - `average` (`fdc`): Average signal
    - `background` (`fbg`): Background signal

- `phasor`: Phasor coordinates
    - `real` (`re`): Real component of phasor coordinates
    - `imag` (`im`): Imaginary component of phasor coordinates

- `polar`: Polar coordinates
    - `phase` (`phi`): Angular component of polar coordinates in radians
    - `modulation` (`mod`): Radial component of polar coordinates

- `lifetime` (`tau`): Fluorescence lifetime components
    - `fraction` (`frac`): Fractional intensity of lifetime component
    - `amplitude` (`amp`): Preexponential amplitude of lifetime component

- `apparent_lifetime` Apparent single lifetime from phase or modulation
    - `phase_lifetime` (`tauphi`) Apparent single lifetime from phase
    - `modulation_lifetime` (`taumod`) Apparent single lifetime from modulation

- `reference`: Signal, coordinates, lifetimes, etc used for calibration
    - `real0` (`re0`): Real component of phasor coordinates for calibration
    - `imag0` (`im0`):  Imaginary component of phasor coordinates for calibration
    - `phase0` (`phi0`):  Angular component of polar coordinates for calibration
    - `modulation0` (`mod0`):  Radial component of polar coordinates for calibration

- `frequency`: Laser or modulation frequency
    - `omega` or `ohm`: Radial frequency
    - `omega_tau`: Radial frequency multiplied with lifetime

I'm not sure how/if to distinguish between un-referenced/corrected and referenced/corrected coordinates. In particular for general polar coordinates vs "phase delay" and "demodulation". At least it should be mentioned in the docstrings what is handled.

Apparent single lifetimes and other conversions are not relevant for this PR, but should be supported by the library at some point:
```Python
def polar_from_reference(
    measured_phase: ArrayLike,
    measured_modulation: ArrayLike,
    known_phase: ArrayLike,
    known_modulation: ArrayLike,
    /,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return polar coordinates from reference measurement."""
    phi0 = measured_phase - known_phase
    mod0 = measured_modulation / known_modulation
    return phi0, mod0


def phasor_from_polar(
    phase: ArrayLike,
    modulation: ArrayLike,
    /,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return phasor coordinates from polar coordinates."""
    ...
    return real, imag


def phasor_from_apparent_lifetimes(
    phase_lifetime: ArrayLike,
    modulation_lifetime: ArrayLike,
    /,
    frequency: ArrayLike,
    *,
    frequency_axis: int | None = None,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return phasor coordinates from apparent single lifetimes."""
    ...
    return real, imag


def phasor_to_apparent_lifetimes(
    real: ArrayLike,
    imag: ArrayLike,
    /,
    frequency: ArrayLike,
    *,
    frequency_axis: int | None = None,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return apparent single lifetimes from phasor coordinates."""
    ...
    return tauphi, taumod


def polar_from_apparent_lifetimes(
    phase_lifetime: ArrayLike,
    modulation_lifetime: ArrayLike,
    /,
    frequency: ArrayLike,
    *,
    frequency_axis: int | None = None,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return polar coordinates from apparent single lifetimes."""
    ...
    return phase, modulation


def polar_to_apparent_lifetimes(
    phase: ArrayLike,
    modulation: ArrayLike,
    /,
    frequency: ArrayLike,
    *,
    frequency_axis: int | None = None,
) -> tuple[NDArray[Any], NDArray[Any]]:
    """Return apparent single lifetimes from polar coordinates."""
    ...
    return tauphi, taumod
```

cgohlke commented:
Below are a simple implementation and tests for the `phasor_from_lifetime` function accepting multiple lifetime components and fractions. It should be good enough for calibration purposes. You can either copy it into your PR, or I can push the changes.

This function should later be extended to optionally accept arrays for frequency, lifetime, and fraction, and return multi-dimensional arrays. That probably needs to be implemented in Cython.

```Python
def phasor_from_lifetime(
    frequency: float,
    lifetime: ArrayLike,
    /,
    fraction: ArrayLike | None = None,
    *,
    is_preexp: bool = False,
) -> tuple[NDArray[numpy.float64], NDArray[numpy.float64]]:
    """Return phasor coordinates from multi-exponential lifetime components.

    Parameters
    ----------
    frequency: float
        Laser pulse or modulation frequency in MHz.
    lifetime: array_like
        Lifetime components in ns. May be a scalar for single-exponential
        or a sequence for multi-exponential components.
    fraction: array_like, optional
        Fractional intensities or pre-exponential amplitudes of the lifetime
        components. Must be of same shape as `lifetime`.
        If None (default), fractions are 1 for all lifetime components.
        Fractions are normalized internally.
    is_preexp: bool, optional
        If true, `fraction` values are pre-exponential amplitudes,
        else fractional intensities (default).

    Returns
    -------
    real, imag: ndarray
        Real and imaginary components of phasor coordinates.

    Raises
    ------
    ValueError
        Input arrays do not match or are more than one dimensional.

    Examples
    --------
    >>> phasor_from_lifetime(80.0, 1.9894368)  # 1000/(2*pi*80) = 1.9894368
    (0..., 0...)
    >>> phasor_from_lifetime(80.0, [3.9788735, 0.9947183], [0.5, 0.4])
    (0..., 0...)
    >>> phasor_from_lifetime(
    ...     80.0, [3.9788735, 0.9947183], [0.5, 0.5], is_preexp=True
    ... )
    (0..., 0...)

    """
    tau = numpy.array(lifetime, dtype=numpy.float64, copy=True)
    if tau.ndim > 1:
        raise ValueError('lifetime must be scalar or one-dimensional array')
    if fraction is None:
        frac = numpy.ones_like(tau)
    else:
        frac = numpy.array(fraction, dtype=numpy.float64, copy=True)
        if tau.shape != frac.shape:
            raise ValueError(f'shape mismatch {tau.shape} != {frac.shape}')
    if is_preexp:
        # preexponential amplitudes to fractional intensities
        frac *= tau
    frac /= numpy.sum(frac)  # TODO: check for zero
    tau *= frequency * 2e-3 * math.pi  # omega_tau
    tmp = numpy.square(tau)
    tmp += 1.0
    tmp **= -1
    tmp *= frac
    re = numpy.sum(tmp)
    tmp *= tau
    im = numpy.sum(tmp)
    return re, im
```
```Python
@pytest.mark.parametrize(
    'args, kwargs, expected',
    [
        # single lifetime
        ((80.0, 0.0), {}, (1.0, 0.0)),  # shortest lifetime
        ((80.0, 1e9), {}, (0.0, 0.0)),  # very long lifetime
        ((80.0, 3.9788735), {}, (0.2, 0.4)),
        ((80.0, 0.9947183), {}, (0.8, 0.4)),
        ((80.0, 1.9894368), {}, (0.5, 0.5)),
        ((80.0, 1.9894368, 1.0), {}, (0.5, 0.5)),
        ((80.0, 1.9894368, 0.1), {}, (0.5, 0.5)),
        ((80.0, [1.9894368]), {}, (0.5, 0.5)),
        ((80.0, [1.9894368], [1.0]), {}, (0.5, 0.5)),
        ((80.0, [1.9894368], [0.6]), {}, (0.5, 0.5)),
        # two lifetimes
        ((80.0, [0.0, 1e9], [0.5, 0.5]), {}, (0.5, 0.0)),
        ((80.0, [0.0, 1e9], [0.6, 0.4]), {}, (0.6, 0.0)),
        ((80.0, [3.9788735, 0.9947183], [0.0, 1.0]), {}, (0.8, 0.4)),
        ((80.0, [3.9788735, 0.9947183], [1.0, 0.0]), {}, (0.2, 0.4)),
        ((80.0, [3.9788735, 0.9947183], [0.5, 0.5]), {}, (0.5, 0.4)),
        ((80.0, [3.9788735, 0.9947183], [0.25, 0.75]), {}, (0.65, 0.4)),
        # three lifetimes
        ((80.0, [0.0, 1.9894368, 1e9]), {}, (0.5, 0.5 / 3)),
        ((80.0, [0.0, 1.9894368, 1e9], [1, 1, 1]), {}, (0.5, 0.5 / 3)),
        ((80.0, [0.0, 1.9894368, 1e9], [0, 1, 0]), {}, (0.5, 0.5)),
        ((80.0, [0.0, 1.9894368, 1e9], [1, 1, 0]), {}, (0.75, 0.5 / 2)),
        ((80.0, [0.0, 1.9894368, 1e9], [0, 1, 1]), {}, (0.25, 0.5 / 2)),
        ((80.0, [0.0, 1.9894368, 1e9], [0, 0, 0]), {}, (numpy.nan, numpy.nan)),
        (
            (80.0, [3.9788735, 1.9894368, 0.9947183], [1.0, 1.0, 1.0]),
            {},
            (0.5, (0.4 + 0.5 + 0.4) / 3),
        ),
        # preexponential amplitudes
        ((80.0, 0.0), {'is_preexp': True}, (numpy.nan, numpy.nan)),  # ?
        ((80.0, 1e9), {'is_preexp': True}, (0.0, 0.0)),
        ((80.0, 3.9788735), {'is_preexp': True}, (0.2, 0.4)),
        ((80.0, [0.0, 1e9], [0.5, 0.5]), {'is_preexp': True}, (0.0, 0.0)),
        (
            (80.0, [3.9788735, 0.9947183], [0.0, 1.0]),
            {'is_preexp': True},
            (0.8, 0.4),
        ),
        (
            (80.0, [3.9788735, 0.9947183], [1.0, 0.0]),
            {'is_preexp': True},
            (0.2, 0.4),
        ),
        (
            (80.0, [3.9788735, 0.9947183], [0.5, 0.5]),
            {'is_preexp': True},
            (0.32, 0.4),
        ),
        (
            (80.0, [3.9788735, 0.9947183], [0.25, 0.75]),
            {'is_preexp': True},
            (0.457143, 0.4),
        ),
    ],
)
def test_phasor_from_lifetime(args, kwargs, expected):
    """Test `phasor_from_lifetime` function."""
    for actual, desired in zip(
        phasor_from_lifetime(*args, **kwargs), expected
    ):
        assert_allclose(actual, desired, atol=1e-6)


def test_phasor_from_lifetime_exceptions():
    """Test exceptions in `phasor_from_lifetime` function."""
    with pytest.raises(ValueError):
        phasor_from_lifetime(80.0, [[0.0]])
    with pytest.raises(ValueError):
        phasor_from_lifetime(80.0, 0.0, [])
    with pytest.raises(ValueError):
        phasor_from_lifetime(80.0, [0.0, 1e9], [1.0])


def test_phasor_from_lifetime_modify():
    """Test `phasor_from_lifetime` function does not modify input."""
    frequency = 80.0
    lifetime = numpy.array([0.0, 1.9894368, 1e9], dtype=numpy.float64)
    fraction = numpy.array([1.0, 1.0, 1.0], dtype=numpy.float64)
    real, imag = phasor_from_lifetime(frequency, lifetime, fraction=fraction)
    assert_allclose(real, 0.5)
    assert_allclose(imag, 0.5 / 3)
    assert_array_equal(frequency, 80.0)  # for future revisions
    assert_array_equal(lifetime, [0.0, 1.9894368, 1e9])
    assert_array_equal(fraction, [1.0, 1.0, 1.0])
```

bruno-pannunzio commented:
Thanks @cgohlke for the implementation of that function, it works great. I updated with the functions we have been discussing and moved them to the `phasor` module. 

Regarding what we talked yesterday, I think I did not understand correctly what the `real0` and `imag0` arguments from the `phasor_from_reference` function you proposed were.  I thought they were the difference between the measured and known real and imaginary from the reference. If that is the case, I don't know how to perform the calibration with that information, as I think it will be needed the original coordinates if we consider the logic I implemented, but maybe there is another way that I am not aware that we can use that to perform the calculation. 

Either way, the polar_from_reference_phasor accepts phasor coordinates and returns the phase and modulation parameters needed for calibration. Those two values (the same as `real0` and `imag0` would be used only for calibration purposes I think. So I don't know if it will matter much if that parameters were in polar or phasor systems, as either of them can be obtained directly from the phasor coordinates without asking the user to convert them to polar. 

Please let me know what you think.

PD: I am seeing some tests are failing because some differences in the decimals of the expected values in tests. I will look into it.

bruno-pannunzio commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,507 @@
+    phase0: ArrayLike = 0.0,
+    modulation0: ArrayLike = 1.0,
+    /,
+) -> tuple[ArrayLike, ArrayLike]:
```
I had to change it to ArrayLike because MyPy was returning an error because of this part:

`    if numpy.all(phi0 == 0) and numpy.all(mod0 == 1):
        return real, imag`

This way if the input is to remain unchanged, the returned elements are exactly the same as the input. I don't know if it is the best way to  go. If you consider better to always return the same type I can force the conversion to numpy array.

cgohlke replied:
MyPy is right. But I think `phi0 ` and `mod0` should be converted to arrays before passing them to `numpy.all(... == 0)` anyway. Also, `phi0==0` and `mod0==1` is an exceptional case, which might not be worth?

bruno-pannunzio commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,507 @@
+    Examples
+    --------
+    >>> phasor_from_lifetime(80.0, 1.9894368)  # 1000/(2*pi*80) = 1.9894368
+    (0..., 0...)
```
Yes I agree. What was easiest was to use values that give more "round" returned values so that we don't have that problem for decimal precision, but I know this might not be possible in all cases. I did a quick research but couldn't find anything for this problem. Maybe we can add an Issue so that we can remember to try find a solution later?

cgohlke replied:
Well, in this case the expected values are `(0.5, 0.5)`. It does not get much rounder than that. Unfortunately the returned values are `0.49999999` or similar due to floating point processing... One option would be to mark the examples with `# doctest: +SKIP`, another option to explicitly round or print the returned values with a certain precision. Both options are not nice for docstring examples. Maybe setting [numpy.set_printoptions](https://numpy.org/doc/stable/reference/generated/numpy.set_printoptions.html) works?

cgohlke replied:
We have to find a solution for this sooner than later.

cgohlke replied:
Turned out I misunderstood the [pytest documentation](https://docs.pytest.org/en/7.1.x/how-to/doctest.html#using-doctest-options). Marking the examples with `# doctest: +NUMBER` works as we want:
```
    >>> phasor_from_lifetime(80.0, 1.9894368)  # doctest: +NUMBER
    (0.5, 0.5)
```

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,507 @@
+    phi0, mod0 = numpy.asarray(phase0), numpy.asarray(modulation0)
+    if numpy.all(phi0 == 0) and numpy.all(mod0 == 1):
+        return real, imag
+    real = numpy.asarray(real)
+    imag = numpy.asarray(imag)
+    correction_matrix = (
+        numpy.array(
+            (
+                (numpy.cos(phi0), -numpy.sin(phi0)),
+                (numpy.sin(phi0), numpy.cos(phi0)),
+            )
+        )
+        * mod0
+    )
+    real, imag = (
+        correction_matrix.dot(numpy.vstack([real.flatten(), imag.flatten()]))
+    ).reshape((2, *real.shape))
```
In the most general case one would want to be able to calibrate every item/pixel of the polar coordinate arrays separately with different reference coordinates. The instrumentation might require that or the phasor coordinates might have a frequency axis. Also, this library is not only intended to process measured images, but also to generate/process synthetic data.

cgohlke replied:
Looking at this implementation, the other functions, and their default use cases, we might want to consider passing/returning coordinate components as shaped `(2, ...)` arrays instead of separate arrays. Internally, that should be simpler and more efficient, but what about users interacting with the library? Thoughts?

This function signature would would simplify to:
```Python
def phasor_calibration(
    phasor: ArrayLike,
    polar0: ArrayLike = (0.0, 1.0)
    /,
) -> NDArray[Any]:
    """Return calibrated/referenced phasor coordinates.

    Parameters
    ----------
    phasor : array_like
        Phasor coordinates to be corrected.
    polar0 : array_like, optional
        Phase and modulation for calibration.

    Returns
    -------
    phasor: ndarray
        Calibrated phasor coordinates.

    Examples
    --------
    >>> phasor = numpy.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
    >>> phasor_calibration(phasor, (0.5, 2.0))
    array(...)
    """
    phi0, mod0 = _coordinates(polar0)
    phasor = _coordinates(phasor)
    ...
```

bruno-pannunzio replied:
I really like that idea. How will the `coordinates` work? Is it like a private helper function that unravels the different components?

I am still trying to make it work with phi0 and mod0 as arrays, so I will update later if I find the solution.

cgohlke replied:
Hold on, my comment was intended to start a discussion, not to change the implementation at this point. There are pro and cons for the approaches...

cgohlke replied:
If it is too complicated to implement some cases using numpy array computing, raise a NotImplementedError or ValueError in those cases for now. Similar to what is in `phasor_from_lifetime`. I think sooner than later we will have to start using Cython to implement some of the low level conversions for efficiency and generalization. But it should probably not delay this PR.

bruno-pannunzio commented:
@cgohlke I updated the`phasor_calibrate` function and now workds with `phase0` and `modulation0` as arrays of the same shape as `real` and `imag`, and added a test to cover for this case. I found a way to implement it that is much more simpler and doesn't require `numpy.vstack`. Is there another case that you think should be covered? For example can `phase0` and `modulation0` be of shape different than `real` and `imag`? 

I also removed from the center function the implemention to `skip_axes` because I found that it did not worked as expected. Can I have some example of when can this be of use? Like the input arrays and the expected output for the skipped axes?

bruno-pannunzio commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,507 @@
+    imag = numpy.asarray(imag)
+    if skip_axes is not None:
+        real = numpy.delete(real, skip_axes, axis=None)
+        imag = numpy.delete(imag, skip_axes, axis=None)
```
This is the part of the code that I removed because it did not have the behaviour I expected. I would need some examples of use cases to test how to implement it.

cgohlke commented:
> Is there another case that you think should be covered? For example can phase0 and modulation0 be of shape different than real and imag? EDIT: it works with phi0 and mod0 of shape different than real and imag because of broadcasting.

> I also removed from the center function the implemention to skip_axes because I found that it did not worked as expected. Can I have some example of when can this be of use? Like the input arrays and the expected output for the skipped axes?

These are related. In addition to these common cases (calibrating phasor coordinate arrays with single polar coordinates or polar coordinates of the same shape as the phasor coordinates), I think we should try to support the case where one or more axes/dimensions (such as frequency or channel) require different coordinates for calibration. Say a user has data from a multi-frequency FLIM measurement and the polar coordinates returned from `polar_from_signal` have the dimensions/axes `(frequency, Y, X)`. Then the `phasor_center` function should be instructed to return center coordinates for each frequency by including the frequency axis in `skip_axes`. And the `phasor_calibrate` should be calibrated with polar coordinates that differ for the frequency axis. Anyway, this might work already thanks to numpy broadcasting, but it should be tested explicitly since we might want to change the implementation later on.

bruno-pannunzio commented on src/phasorpy/phasor.py:
```diff
@@ -57,50 +56,47 @@ def phasor_calibration(
-    ).reshape((2, *real.shape))
-    return real, imag
+        return re, im
+    re_calibrated = (re * numpy.cos(phi0) - im * numpy.sin(phi0)) * mod0
+    im_calibrated = (re * numpy.sin(phi0) + im * numpy.cos(phi0)) * mod0
+    return re_calibrated, im_calibrated
```
Thanks Christoph, It is more efficent that way. Can I ask about the modification in:
` re = numpy.array(real, copy=True)
    im = numpy.array(imag, copy=True)`
    
Is there a reason to always create a copy?

cgohlke replied:
A copy is needed because of the in-place modification `im *= cos` and `im_calibrated *= sin`. In fact, it should be verified in the tests that the input data is not modified.

bruno-pannunzio replied:
You are right, I will add tests to cover this aspect

bruno-pannunzio replied:
@cgohlke I had to add dtype convertion in the `numpy.array` to float because of casting error with int and float, in the case were real and imag were `int`. I think those cases won't be used too much, but anyway, with this fix it works for all cases:

`    re = numpy.array(real, copy=True, dtype=float)
    im = numpy.array(imag, copy=True, dtype=float)`

Do you think this can alter the process or be less efficent?

cgohlke replied:
Yes, we might want to do that for all input coordinates and lifetimes in all functions. One issue is to keep `float32` input as is (to save memory for large images). I see no other way than have a `dtype` input parameter that defaults to `float64` but can be set to `float32`.

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -1,28 +1,27 @@
+    fluorescence lifetime
+- reduce arrays of phasor coordinates to single coordinates
 
 """
```
`reference signal of known fluorescence lifetime` -> `reference of known fluorescence lifetime`?

bruno-pannunzio replied:
Yes I will modify that line since the input of reference can be both polar or signal.

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,511 @@
+
+    Examples
+    --------
+    >>> real = numpy.array([1.0, 2.0, 3.0])
+    >>> imag = numpy.array([4.0, 5.0, 6.0])
+    >>> phase0 = 0.5
+    >>> modulation0 = 2.0
+    >>> real_calibrated, imag_calibrated = phasor_calibrate(
+    ...     real, imag,
+    ...     phase0,
+    ...     modulation0
+    ... )
+    >>> real_calibrated
+    array(...)
+    >>> imag_calibrated
+    array(...)
```
I find these examples quite verbose for docstrings. The function signature is explained just a few lines before and these docstrings are often displayed by IDEs in small  popup windows.

How about:
```
Examples
--------
Use scalar reference coordinates to calibrate phasor coordinates:
>>> phasor_calibrate([1.0, 2.0, 3.0], [4.0, 5.0, 6.0], 0.5, 0.2)
(array(...), array(...))

Use separate reference coordinates for each phasor coordinate:
>>> phasor_calibrate([1.0, 2.0, 3.0], [4.0, 5.0, 6.0], [0.5, 0.2, 0.3], [1.5, 2.0, 0.3])
(array(...), array(...))
```
What do you think? Maybe get some feedback from Bruno and Leonel?

bruno-pannunzio replied:
Bruno and Leonel also agreed that maybe the less verbose and more concise example are better. So we can start using that criteria from now on.

cgohlke replied:
Good. We can always improve/revise/add later.

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,511 @@
+    imag : array_like
+        Scalar or array containing the imaginary components of the
+        phasor coordinates.
+    skip_axes : tuple[int, ...] or None, optional
```
`skip_axes : tuple of int, optional`

cgohlke commented:
> I had to remove the geometric_median method because I could find an easy solution to make it work with skip_axes without having to import some libraries for optimization like scipy.

You could raise a ValueError or NotImplementedError in those cases. We already discussed that it is OK to depend on Scipy, scikit-image, or scikit-learn, but their import should be deferred.

cgohlke commented:
At this point the PR could benefit from the input of @schutyb and @lmalacrida about the proposed terminology, API, and the style and verbosity of the docstrings.

I still find the proposed workflow a bit odd, having to switch back and forth between phasor and polar coordinates:
```Python
signal = read_format('signal.format')
reference_signal = read_format('reference.format')
reference_lifetime = 4.0  # may be extracted from `reference_signal.attrs`
frequency = 80.0  # may be extracted from `signal.attrs`
axis = 0  # may be determined from `signal.attrs`

phasor = phasor_calibrate(
    *phasor_from_signal(signal, axis=axis)[:2],
    *polar_from_reference_phasor(
        *phasor_center(
            *phasor_from_signal(reference_signal, axis=axis)[:2],
            method='mean',
        ),
        *phasor_from_lifetime(frequency, reference_lifetime)
    ),
) 
```
Anyway, we need an implementation of `phasor_from_signal` to start writing a tutorial about this functionality.

bruno-pannunzio commented:
I agree, maybe we can pause the iterations of this PR until we are all in the same track regarding those central issues of this library. We can make a list of this issues to be discussed in the next meeting.

cgohlke commented:
I don't think this PR and the discussion should wait for another 10 days to move on. If Leonel and Bruno can't comment here, maybe you can sit down with them to get feedback on the following:

- terminology
- API, function names, parameter names
- proposed workflow for calibration (standard case)
- docstring style and verbosity

Do we still consider passing coordinates as `(2, ...)` shaped arrays instead of separate arrays?

bruno-pannunzio commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,511 @@
+    /,
+) -> tuple[NDArray[Any], NDArray[Any]]:
+    """Return calibrated/referenced phasor coordinates.
+
```
Do you think this is too long? Maybe omit the last part?

    ```
     Return calibrated/referenced phasor coordinates.

    Calibration of phasor coordinates in fluorescence lifetime analysis is 
    necessary to account for the instrument response function (IRF) and delays
    in the electronics.
    
    This function can also be used to transform/rotate any phasor coordinate.
```

cgohlke replied:
That's good.

bruno-pannunzio commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,511 @@
+
+    Parameters
+    ----------
+    real : array_like
+        Scalar or array containing the real components of phasor coordinates.
+    imag : array_like
+        Scalar or array containing the imaginary components of phasor
+        coordinates.
+    phase0 : array_like, optional
+        Angular component of polar coordinates for calibration. Defaults
+        to 0.0.
+    modulation0 : array_like, optional
+        Radial component of polar coordinates for calibration. Defaults to 1.0.
```
I like the idea to make it cleaner but I tried in Sphinx documentation and I think it doesn't work really well as expected:
<img width="608" alt="image" src="https://github.com/phasorpy/phasorpy/assets/90658672/b039ca92-2612-47c8-84a3-512103e4fbbb">

cgohlke replied:
Too bad. The not. It seems to work for return values (?)

cgohlke replied:
Try to include the return type, something like:
```
    Returns
    -------
    phase0, modulation0: ndarray
        Angular and radial components of polar coordinates for calibration.
```

cgohlke replied:
Or we just keep the returns description like/symmetric with the parameter description even if it is a bit verbose...

bruno-pannunzio replied:
Thanks Christoph! Including the type fixed it and now it displays correctly:
<img width="610" alt="image" src="https://github.com/phasorpy/phasorpy/assets/90658672/4f172748-dd99-4567-94cc-c500dccfcd33">

I feel this works although returns and parameters don't share the same descriptions. But if you prefer to have the same I think it also works well

cgohlke replied:
Looks a bit confusing. Let's keep returns and parameters the same.

bruno-pannunzio commented:
@cgohlke we talked today with Bruno and Leonel about the topics you listed. Leonel will comment later on some aspects of the terminology and proposed names. But I believe that they don't directly affect this PR. 

They also believe function names and proposed workflow for calibration standard are correct.

> Do we still consider passing coordinates as (2, ...) shaped arrays instead of separate arrays?

I think for now we can have separate arrays and maybe think of implementing it later? Or that would set back too much later?

Also regarding the convertion of dtype to float32 or float64. Do you think I should implement it for all function in this PR or leave it for a future PR?

I commited the changes in docstring we have been discussing for review.

cgohlke commented:
>> Do we still consider passing coordinates as (2, ...) shaped arrays instead of separate arrays?

> I think for now we can have separate arrays and maybe think of implementing it later? Or that would set back too much later?

All options have pros and cons. Given that #13 also uses separate arrays, let's keep it that way.

cgohlke commented on docs/api/index.rst:
```diff
@@ -17,3 +17,4 @@ PhasorPy library version |version|.
     datasets
     io
     cli
+    phasor
```
Let's make the phasor module more prominent and change the order to:
```
    phasorpy
    phasor
    io
    color
    datasets
    cli
```

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,509 @@
+
+    Examples
+    --------
+    Use scalar reference coordinates to calibrate phasor coordinates:
```
This does not render correctly. It needs a blank line after:
```
    Examples
    --------
    Use scalar reference coordinates to calibrate phasor coordinates:

    >>> phasor_calibrate([1.0, 2.0, 3.0], [4.0, 5.0, 6.0], 0.5, 0.2)
    (array(...), array(...))

    Use separate reference coordinates for each phasor coordinate:

    >>> phasor_calibrate(
    ... [1.0, 2.0, 3.0], [4.0, 5.0, 6.0], [0.5, 0.2, 0.3], [1.5, 2.0, 0.3]
    ... )
    (array(...), array(...))
```

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,509 @@
+    known_imag: ArrayLike,
+    /,
+) -> tuple[NDArray[Any], NDArray[Any]]:
+    """Return  components for calibration from reference phasor.
```
`"""Return polar coordinates for calibration from reference phasor.`

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,509 @@
+        Angular component of polar coordinates for calibration.
+    modulation0: ndarray
+        Radial component of polar coordinates for calibration.
+    Raises
```
Add blank line before.

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,509 @@
+    --------
+    >>> polar_from_reference_phasor(0.5, 0.0, 1.0, 0.0)
+    (0.0, 2.0)
+    """
```
Docstrings should end with a blank line.

lmalacrida commented:
@cgohlke today we discussed with @bruno-pannunzio @schutyb about the following points:

- terminology

- API, function names, parameter names
- 
These are my insights on the following terminology and function names:


signal: Time-domain, frequency-domain, or spectral data

decay, histogram: time-domain signal
wave: Frequency-domain signal
spectrum: Spectral signal
average (fdc): Average signal
background (fbg): Background signal
phasor: Phasor coordinates

**- I confess that I dislike the term wave; while I understand from where it comes from, nowadays, we recover a phase histogram to perform the lifetime on a digital-frequency domain. Then, I believe both signals from the time and frequency domain are histograms, the first one from time bins and the second from phase bins. Why not call it as: Time, histogram and Phase, histogram?
- The others are fine with me.** 

real (re): Real component of phasor coordinates
imag (im): Imaginary component of phasor coordinates
polar: Polar coordinates

- Here, we also discussed if real and imag should be the proper names. While I agree these are general terms of the Fourier transform, Weber named them G and S, and I believe we should stick to them. We can specifically discuss in our next meeting. 

phase (phi): Angular component of polar coordinates in radians
modulation (mod): Radial component of polar coordinates
lifetime (tau): Fluorescence lifetime components

- Fine

fraction (frac): Fractional intensity of lifetime component
amplitude (amp): Preexponential amplitude of lifetime component
apparent_lifetime Apparent single lifetime from phase or modulation

- Here, we should also define a function correlating the fractional intensity and the preexponential factors. 
- I also want to discuss if it should be called the Fractional intensity of the Lifetime component or the Fraction of photons of lifetime components. At the end, we are measuring photons.
- 

phase_lifetime (tauphi) Apparent single lifetime from phase
modulation_lifetime (taumod) Apparent single lifetime from modulation
reference: Signal, coordinates, lifetimes, etc used for calibration

-Fine

real0 (re0): Real component of phasor coordinates for calibration
imag0 (im0): Imaginary component of phasor coordinates for calibration
phase0 (phi0): Angular component of polar coordinates for calibration
modulation0 (mod0): Radial component of polar coordinates for calibration
frequency: Laser or modulation frequency

- Fine, we should correct in regard to what we decided with G/S or re/img

omega or ohm: Radial frequency
omega_tau: Radial frequency multiplied with lifetime

- fine
I believe it is important to discuss some of these things to avoid progressing with terminology we can decide to change afterward.

cgohlke commented:
> I confess that I dislike the term wave; while I understand from where it comes from, nowadays, we recover a phase histogram to perform the lifetime on a digital-frequency domain. Then, I believe both signals from the time and frequency domain are histograms, the first one from time bins and the second from phase bins. Why not call it as: Time, histogram and Phase, histogram?

It's not really relevant for this PR. On the one hand, I don't think of homodyne frequency-domain data, for example, full-field images taken at certain phase shifts, as "histograms". Also, simply referring to "histogram" might cause confusion with the histogram of the phasor coordinates. That's why I prefer `signal`. On the other hand, phase images will be treated by the library just the same as time-domain and digital frequency-domain histograms. Also, the `phasorpy.io` module standardized on the `H` (histogram) dimension for returning the phase images. At some point we will need a way to distinguish histograms/signals from time-domain and frequency-domain in function names. In the past I used `decay` and `wave` but we could also use the full `time_domain_signal/histogram`  and `frequency_domain_signal/histogram` names, which is probably less confusing.

> Here, we also discussed if real and imag should be the proper names. While I agree these are general terms of the Fourier transform, Weber named them G and S, and I believe we should stick to them. We can specifically discuss in our next meeting.

Having used `g` and `s` myself in the past, I'm opposed to single letter parameter names in function signatures or even as variable names unless for loop indices (`i`, `j`, `k`), short lived variables, or trivial cases. All-uppercase names are reserved for module constants by convention in Python. `real` and `imag` are used by numpy and are recognized by a wider audience than `G` and `S`. The documentation could certainly reference `G` and `S` by Weber. If using `G` and `S`, should we also use `g`, `s`, `a`, `α`, `τ`, `φ`, `ω` etc as variable names?

>  also want to discuss if it should be called the Fractional intensity of the Lifetime component or the Fraction of photons of lifetime components. At the end, we are measuring photons.
> Here, we should also define a function correlating the fractional intensity and the preexponential factors.

"Fractional intensity" and "pre-exponential amplitude," are common terms, which are also used in your review article I see. The relationship between those two and also to the photon counts can certainly be added to the documentation. So far we have not integrated math formulas or theory sections in the documentation. I think it is good enough to reference publications in most cases. Btw, the `phasor_from_lifetime` function is still unfinished, but appropriate for this PR.

cgohlke commented:
> While I agree these are general terms of the Fourier transform, Weber named them G and S, and I believe we should stick to them.

Yet Clegg used `X` and `Y`, Clayton `A` and `B`... All of these names are non-descriptive. If you must use `G` and `S`, use lower case `g` and `s` names, make parameters positional only, and keep the current descriptions, for example `g: real component of phasor coordinates` so at least other groups can relate.

cgohlke commented:
Hi @bruno-pannunzio 

I was tikering with docstring examples and found the default numpy array print options too verbose. How about we limit precision and number of array items printed in docstrings using the following pytest fixture in `src/phasorpy/conftest.py`? Can you try to integrate it in this PR?

```Python
@pytest.fixture(autouse=True)
def set_printoptions():
    """Adjust numpy array print options for use with `# doctest: +NUMBER`."""
    numpy.set_printoptions(
        # precision=3,
        threshold=5,
        formatter={'float': lambda x: f'{x:.4g}'},  # remove whitespace
    )
```

I noticed that there is not yet an agreement on units. In some cases the `io` module returns values in different units than expected by this PR. How about:

- `MHz` for laser/modulation frequencies
- `ns` for fluorescence lifetimes
- `ns` for TCSPC histograms (or `ps`?)
- `nm` for wavelengths
- `m` for spatial coordinates
- `radian` for angles

Or should we consistently use SI units without prefix, that is `Hz`, `s`, `m`? Thoughts?

I'm working on replacing the `phasor_from_lifetime` function with a more capable implementation in Cython. Do you think this PR can be finished soon so I can build on it?

bruno-pannunzio commented:
@cgohlke I tried adding the pytest fixture and changed the examples in the docstrings. I am having this error with pytest:

```
Expected:
    (array([-0.20802392, -0.12839251, -0.04876111]), array([0.79795116, 1.06935278, 1.3407544 ]))
Got:
    (array([-0.208, -0.1284, -0.04876]), array([0.798, 1.069, 1.341]))
```

Is it necessary to round floats to 4 decimals in the docstrings with that fixture?

Regarding the units. I believe its better to go with the ones you proposed:

> MHz for laser/modulation frequencies
> ns for fluorescence lifetimes
> ns for TCSPC histograms (or ps?)
> nm for wavelengths
> m for spatial coordinates
> radian for angles

Let me ask Leonel and I will let you know.

I believe we can close this PR between today and tomorrow.

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -331,13 +324,13 @@ def phasor_from_lifetime(
     ...     80.0, [3.9788735, 0.9947183], [0.5, 0.5], is_preexp=True
     ... )
-    (0..., 0...)
+    (0.3200000038038807, 0.39999999917836626)
```
Mark with `# doctest: +NUMBER`, then it simplifies to `(0.32, 0.4)`

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -279,7 +270,9 @@ def phasor_to_polar(
     ...     [1.0, 2.0, 3.0], [4.0, 5.0, 6.0]
     ... )
-    (array(...), array(...))
+    (array([1.326, 1.19, 1.107]),
+    array([4.123, 5.385, 6.708]))
```
Could be on one line? `(array([1.326, 1.19, 1.107]), array([4.123, 5.385, 6.708]))`

cgohlke commented on src/phasorpy/conftest.py:
```diff
@@ -9,3 +10,12 @@
 def add_fetch(doctest_namespace):
     """Add datasets.fetch to doctest namespace."""
     doctest_namespace['fetch'] = fetch
+    
```
Two lines between top level function definitions.

cgohlke commented on src/phasorpy/phasor.py:
```diff
@@ -0,0 +1,514 @@
+
+    Examples
+    --------
+    Compute center coordinates with the 'mean' method:
+    >>> phasor_center([1.0, 2.0, 3.0], [4.0, 5.0, 6.0], method='mean')
+    (2.0, 5.0)
+
+    Compute center coordinates with the 'median' method:
+    >>> phasor_center([1.0, 2.0, 3.0], [4.0, 5.0, 6.0], method='median')
+    (2.0, 5.0)
```
These examples still don't render correctly. Add an empty line between text and code.

cgohlke commented:
Nice! I think this can be merged. I haven't looked at all the test cases but coverage is 100% and it looks in great shape.

TODO list from this discussion:

- [ ] Leave a comment in #13 about which functions were implemented here.
- [ ] Discuss and decide on naming components of phasor coordinate g/s, real/imag, or re/im.
- [ ] Special handle `float32` input?
- [ ] Update `phasorpy.io` to follow metadata naming and units used in this PR.
- [ ] Replace `phasor_from_lifetime` function with more capable one.
- [ ] Implement missing functions: `phasor_from_signal`, `phasor_from_apparent_lifetime`, `phasor_to_apparent_lifetime`, `phasor_from_polar`
- [ ] Write a tutorial once `phasor_from_signal` is implemented.

bruno-pannunzio commented:
Thanks @cgohlke for the TODO list, this is very usefull to keep track of the improvements needed. I would include also adding other methods to `phasor_center` like 'kmeans' and 'geometric_median'.

Should we create an issue/issues to make more visible so that maybe other contributors can tackle this topics?

cgohlke commented:
There's #21 for a general TODO list but it's rather unspecific. If you like, copy or reference this TODO list over there?

`phasor_from_signal` is #20.

`phasor_from_lifetime`, `phasor_from_apparent_lifetime`, `phasor_to_apparent_lifetime`, `phasor_from_polar` fit in #6. I'm working on improving the first function.

The `special handle float32 input` issue will likely come up again once we start processing large data and might need to consider re-implementing several functions in Cython.

For now, if you like, open a specific issue for "Implement `'kmeans'` and `'geometric_median'` methods for `phasor_center`".

cgohlke commented:
Closes #27?

lmalacrida commented:
I agree with the unit proposed by @cgohlke, my only doubt is the ps for TCSPC, because I believe most of them use ps.
Let's discuss it next Monday.

cgohlke commented:
Some more thoughts on units: many (most?) file formats store frequency in `Hz`, wavelengths and spatial sizes in `m`, and times in `s`. Of course there are exceptions. Lifetimes are most often in `ns` or `ps` it seems.

The advantage of using SI units without prefix is that conversions will be less often required and slightly easier if required. It's also easier to document and justify. On the other hand, interactive usage and tutorial examples will be slightly less straightforward when requiring no-prefix units. For example `phasor_from_lifetime(80e6, [1.9e-9, 4.2e-9])` vs `phasor_from_lifetime(80, [1.9, 4.2])`. But then, `1.9e-9` shows right away the this is a lifetime in the `ns` range and not `ms` or `ps`...
